### PR TITLE
(core) dms ingestion stage

### PR DESCRIPTION
**Pull request by malachi-constant**
### Feature 
- DMS Ingestion Stage 

### Detail
- DMSFactory
- DMS S3 to S3 Stage (pending validation of functionality)

### Relates
- #89 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/aws_ddk_core/resources/_dms.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    """
+    Class factory create and configure Kinesis DDK resources, including Delivery Streams.
+    """
+
+    @staticmethod
+    def endpoint_settings_s3(
+        scope: Construct,
+        id: str,
+        environment_id: str,
+        bucket_name: str,
```
Comment (kukushking) [inline core/aws_ddk_core/resources/_dms.py:53]: Can we add prefix here as well? In case ingesting into the shared raw bucket

Comment (malachi-constant) [reply core/aws_ddk_core/resources/_dms.py:62]: added `bucket_folder`

File: core/aws_ddk_core/stages/dms_s3_to_s3.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                    ]
+                },
+            )
+            if not table_mappings
+            else table_mappings,
+        )
+
+        event_detail: Dict[str, Any] = {
+            "bucket": {
+                "name": [target_bucket.bucket_name],
```
Comment (kukushking) [inline core/aws_ddk_core/stages/dms_s3_to_s3.py:126]: Should we add prefix here as well?

File: core/aws_ddk_core/resources/_dms.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+class DMSEndpointS3SettingsConfiguration(BaseSchema):
+    """DDK DMS Endpoint S3 Settings Marshmallow schema."""
+
+    enable_statistics = fields.Bool()
+    max_file_size = fields.Int()
+
+
+class DMSFactory:
+    """
+    Class factory create and configure Kinesis DDK resources, including Delivery Streams.
```
Comment (jaidisido) [inline core/aws_ddk_core/resources/_dms.py:54]: Minor - DMS, not Kinesis

File: core/aws_ddk_core/stages/dms_s3_to_s3.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+from typing import Any, Dict, List, Optional, Union
+
+from aws_cdk.aws_events import EventPattern, IRuleTarget
+from aws_cdk.aws_s3 import IBucket
+from aws_ddk_core.pipelines.stage import DataStage
+from aws_ddk_core.resources import DMSFactory
+from constructs import Construct
+
+
+class DMSS3ToS3Stage(DataStage):
```
Comment (kukushking) [inline core/aws_ddk_core/stages/dms_s3_to_s3.py:24]: With support of other integrations besides S3 I'd look into making this an abstract DMS stage that can be configured with different source/target endpoints

**Review by kukushking — APPROVED** [review]

File: core/aws_ddk_core/resources/_dms.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+        environment_id: str
+            Identifier of the environment
+        bucket_name: str
+            The name of the S3 bucket.
+        bucket_folder: Optional[str]
+            An optional parameter to set a folder name in the S3 bucket.
+            If provided, tables are created in the path *bucketFolder* / *schema_name* / *table_name* / .
+            If this parameter isn’t specified, the path used is *schema_name* / *table_name* / .
+        service_access_role_arn: Optional[str]
+            A required parameter that specifies the Amazon Resource Name (ARN) used by
```
Comment (jaidisido) [inline core/aws_ddk_core/resources/_dms.py:92]: A few things on this one:
1. Should we say it's required if the argument is set to optional?
2. We say the role must allow `iam:PassRole` but I don't see us doing that in the logic below?
3. Shall we say that if `None`, a role is created?

File: core/aws_ddk_core/resources/_dms.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        enable_statistics: Optional[bool] = None,
+        max_file_size: Optional[int] = None,
+        **endpoint_s3_props: Any,
+    ) -> dms.CfnEndpoint.S3SettingsProperty:
+        """
+        Create and configure DMS endpoint settings for s3.
+
+        This construct allows to configure parameters of the dms endpoint using ddk.json
+        configuration file depending on the `environment_id` in which the function is used.
+        Supported parameters are: ...
```
Comment (jaidisido) [inline core/aws_ddk_core/resources/_dms.py:75]: enable_statistics and max_file_size

File: core/tests/unit/test_dms.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+
+from aws_cdk.assertions import Template
+from aws_ddk_core.base import BaseStack
+from aws_ddk_core.resources import DMSFactory, S3Factory
+
+
+def test_s3_source_endpoint(test_stack: BaseStack) -> None:
```
Comment (jaidisido) [inline core/tests/unit/test_dms.py:21]: The two tests below are pretty identical, so I would reduce it to one and parameterise with:
```
@pytest.mark.parametrize("type", ["source", "target"])
```

File: core/tests/unit/test_dms.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        "AWS::DMS::Endpoint",
+        props={
+            "EndpointType": "target",
+            "EngineName": "s3",
+            "S3Settings": {"BucketName": {"Ref": "dummybucket12E106EF4"}},
+        },
+    )
+
+
+def test_s3_to_s3_replication(test_stack: BaseStack) -> None:
```
Comment (jaidisido) [inline core/tests/unit/test_dms.py:83]: I would consider adding DMS configuration in the conftest file to check whether resources read ddk config correctly

File: core/aws_ddk_core/stages/dms_s3_to_s3.py (hunk 4 above)
Comment (jaidisido) [inline core/aws_ddk_core/stages/dms_s3_to_s3.py:24]: Nice!

**Review by jaidisido — COMMENTED** [review]
Looks great, left a few comments here and there

File: core/aws_ddk_core/stages/dms_s3_to_s3.py (hunk 2 above)
Comment (malachi-constant) [reply core/aws_ddk_core/stages/dms_s3_to_s3.py:134]: https://github.com/awslabs/aws-ddk/pull/105/commits/4c1719f9a066e259d68bca6956e6d47240494db6

File: core/aws_ddk_core/resources/_dms.py (hunk 5 above)
Comment (malachi-constant) [reply core/aws_ddk_core/resources/_dms.py:92]: 1. Yep fixed, this is not required in our construct.
2. This is not needed, was carried over from my initial dev.
3. Added

File: core/tests/unit/test_dms.py (hunk 7 above)
Comment (malachi-constant) [reply core/tests/unit/test_dms.py]: Awesome suggestion, thanks. TIL

File: core/tests/unit/test_dms.py (hunk 8 above)
Comment (malachi-constant) [reply core/tests/unit/test_dms.py:54]: Agreed, done.

**Review by jaidisido — APPROVED** [review]

**Pull request merged**